cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -0,0 +1,421 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Represents Boolean functions as a series of CNOT and rotation gates. The Boolean functions are
+passed as Sympy expressions and then turned into an optimized set of gates.
+
+References:
+[1] On the representation of Boolean and real functions as Hamiltonians for quantum computing
+    by Stuart Hadfield, https://arxiv.org/pdf/1804.09130.pdf
+[2] https://www.youtube.com/watch?v=AOKM9BkweVU is a useful intro
+[3] https://github.com/rsln-s/IEEE_QW_2020/blob/master/Slides.pdf
+[4] Efficient quantum circuits for diagonal unitaries without ancillas by Jonathan Welch, Daniel
+    Greenbaum, Sarah Mostame, Alán Aspuru-Guzik, https://arxiv.org/abs/1306.3991
+"""
+
+import functools
+import itertools
+from typing import cast, Any, Callable, Dict, Generator, List, Sequence, Tuple
+
+from sympy.logic.boolalg import And, Not, Or, Xor
+from sympy.core.expr import Expr
+from sympy.core.symbol import Symbol
+import sympy.parsing.sympy_parser as sympy_parser
+
+import cirq
+from cirq import value
+from cirq.ops import raw_types
+
+_hamiltonian_O: Callable[[], 'cirq.PauliString'] = lambda: 0.0 * cirq.PauliString({})
+_hamiltonian_I: Callable[[], 'cirq.PauliString'] = lambda: cirq.PauliString({})
+_hamiltonian_Z: Callable[['cirq.Qid'], 'cirq.PauliString'] = lambda qubit: cirq.PauliString(
+    {qubit: cirq.Z}
+)
+
+
+def _build_hamiltonian_from_boolean(
```

Comment:
I think this should be a separate public function in itself that could be split out from this PR. That will make reviewing easier as well.

Reply:
Sent #4282